File: go/canonical.go
package audit

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"unicode/utf16"
)

// CanonicalJSON serializes a decoded JSON value — exactly the value domain
// json.Decoder with UseNumber produces (nil, bool, json.Number, string, []any,
// map[string]any) — to canonical JSON, byte-for-byte identical to the TS
// canonicalJsonStringify and the Rust canonical_json.
//
// Contract (must NOT drift across languages):
//   - primitives  → JSON.stringify semantics (see writeJSString / numbers)
//   - arrays      → "[" + items.join(",") + "]"   (order PRESERVED)
//   - objects     → keys SORTED by UTF-16 code-unit order at every depth, each
//     rendered  "key":value  joined by ","; no insignificant whitespace.
//
// A JSON null that is *present* in the value (e.g. diff.after on a delete)
// renders as "null" — it is NOT omitted. Absent optional fields never reach here:
// the emitter drops them (omitempty) before the value is built, mirroring the TS
// "value is undefined → key omitted" rule.
func CanonicalJSON(v any) (string, error) {
	var b strings.Builder
	if err := writeCanonical(&b, v); err != nil {
		return "", err
	}
	return b.String(), nil
}

func writeCanonical(b *strings.Builder, v any) error {
	switch val := v.(type) {
	case nil:
		b.WriteString("null")
	case bool:
		if val {
			b.WriteString("true")
		} else {
			b.WriteString("false")
		}
	case json.Number:
		// ponytail: source-text passthrough. Matches JS integer rendering (the
		// audit value domain: counts, rates, ids-as-strings). A source exponent
		// form like "1e3" would need JS Number formatting ("1000") — such values
		// are not part of the event schema. Upgrade path: parse + reformat here
		// if a float/exponent field ever lands.
		b.WriteString(val.String())
	case string:
		writeJSString(b, val)
	case []any:
		b.WriteByte('[')
		for i, item := range val {
			if i > 0 {
				b.WriteByte(',')
			}
			if err := writeCanonical(b, item); err != nil {
				return err
			}
		}
		b.WriteByte(']')
	case map[string]any:
		keys := make([]string, 0, len(val))
		for k := range val {
			keys = append(keys, k)
		}
		sort.Slice(keys, func(i, j int) bool { return lessUTF16(keys[i], keys[j]) })
		b.WriteByte('{')
		for i, k := range keys {
			if i > 0 {
				b.WriteByte(',')
			}
			writeJSString(b, k)
			b.WriteByte(':')
			if err := writeCanonical(b, val[k]); err != nil {
				return err
			}
		}
		b.WriteByte('}')
	default:
		return fmt.Errorf("audit: cannot canonicalize value of type %T", v)
	}
	return nil
}

// writeJSString reproduces ECMAScript JSON.stringify string quoting exactly, so
// the bytes match V8 (and the TS/Rust ports) rather than Go's encoding/json.
//
// Go's encoder is NOT usable here: even with SetEscapeHTML(false) it escapes
// U+2028/U+2029 (JSON.stringify does not), and by default escapes <,>,& . This
// escaper: quote + backslash escaped; \b \t \n \f \r short escapes; other
// control chars (<0x20) as lowercase \uXXXX; forward slash NOT escaped; every
// other rune (accented, astral emoji, …) emitted as literal UTF-8 — matching
// JSON.stringify, whose output encoded as UTF-8 is literal for non-ASCII.
func writeJSString(b *strings.Builder, s string) {
	b.WriteByte('"')
	for _, r := range s {
		switch r {
		case '"':
			b.WriteString(`\"`)
		case '\\':
			b.WriteString(`\\`)
		case '\b':
			b.WriteString(`\b`)
		case '\t':
			b.WriteString(`\t`)
		case '\n':
			b.WriteString(`\n`)
		case '\f':
			b.WriteString(`\f`)
		case '\r':
			b.WriteString(`\r`)
		default:
			if r < 0x20 {
				fmt.Fprintf(b, `\u%04x`, r)
			} else {
				b.WriteRune(r)
			}
		}
	}
	b.WriteByte('"')
}

// lessUTF16 compares two strings by UTF-16 code-unit order — the ordering JS
// String comparison (and thus sorting Object.keys) uses. For BMP text this
// equals code-point (and UTF-8 byte) order; it diverges only for astral
// characters, whose surrogate code units (0xD800–0xDBFF) sort BEFORE BMP chars
// ≥ 0xE000. Audit key names are ASCII in practice, but we match JS exactly so an
// arbitrary metadata key can never break cross-language hash parity.
func lessUTF16(a, b string) bool {
	au := utf16.Encode([]rune(a))
	bu := utf16.Encode([]rune(b))
	for i := 0; i < len(au) && i < len(bu); i++ {
		if au[i] != bu[i] {
			return au[i] < bu[i]
		}
	}
	return len(au) < len(bu)
}

File: rust/audit/src/canonical.rs
//! Canonical JSON serialization — a faithful port of the TS `canonicalJsonStringify`
//! (`@smooai/audit`), byte-verified against `spec/parity-corpus.json`.
//!
//! # The contract, restated
//!
//! ```text
//! canonical(value):
//!   - null / primitive → JSON.stringify(value)
//!   - array            → "[" + items.map(canonical).join(",") + "]"   (ORDER preserved)
//!   - object           → keys with a non-`undefined` value, SORTED ascending by
//!                        JS string (UTF-16 code-unit) order, each rendered
//!                        JSON.stringify(key) ":" canonical(value), joined by ",".
//! ```
//!
//! We operate over [`serde_json::Value`]. Absent optionals are already dropped
//! by serde's `skip_serializing_if = "Option::is_none"` before the Value is
//! built (the TS "value is `undefined` → key omitted" rule); a *present* JSON
//! `null` renders as `"null"`, matching `JSON.stringify(null)`.
//!
//! `serde_json::to_string` matches `JSON.stringify` for the audit value domain:
//! `\"`/`\\` escaped, forward slash NOT escaped, control chars as `\uXXXX`,
//! non-ASCII emitted as literal UTF-8, and integers rendered identically. Object
//! keys are our own ASCII camelCase field names, so a byte `sort()` equals JS
//! UTF-16 order — no astral-plane key ever reaches the sort.

use serde_json::Value;

/// Serialize a [`serde_json::Value`] to its canonical JSON string, byte-matching
/// the TS `canonicalJsonStringify`: keys sorted at every depth, array order
/// preserved, no insignificant whitespace.
pub fn canonical_json(value: &Value) -> String {
    match value {
        Value::Null | Value::Bool(_) | Value::Number(_) | Value::String(_) => {
            // Infallible for a primitive Value; the "null" fallback only guards
            // the impossible error path so this never panics on the seal path.
            serde_json::to_string(value).unwrap_or_else(|_| "null".to_string())
        }
        Value::Array(items) => {
            let mut out = String::from("[");
            for (i, item) in items.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                out.push_str(&canonical_json(item));
            }
            out.push(']');
            out
        }
        Value::Object(map) => {
            // Sort keys ascending — matches the TS `.sort()`. ASCII camelCase key
            // bytes == JS UTF-16 order.
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort();
            let mut out = String::from("{");
            for (i, key) in keys.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                let key_json = serde_json::to_string(*key).unwrap_or_else(|_| format!("\"{key}\""));
                out.push_str(&key_json);
                out.push(':');
                out.push_str(&canonical_json(&map[*key]));
            }
            out.push('}');
            out
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn primitives_match_json_stringify() {
        assert_eq!(canonical_json(&json!(null)), "null");
        assert_eq!(canonical_json(&json!(true)), "true");
        assert_eq!(canonical_json(&json!(42)), "42");
        assert_eq!(canonical_json(&json!("a\"b\\c/d\n")), "\"a\\\"b\\\\c/d\\n\"");
    }

    #[test]
    fn objects_sort_arrays_preserve() {
        let v = json!({ "b": 1, "a": { "z": 2, "y": 3 }, "c": [3, 1, 2] });
        assert_eq!(canonical_json(&v), "{\"a\":{\"y\":3,\"z\":2},\"b\":1,\"c\":[3,1,2]}");
    }

    #[test]
    fn present_null_is_rendered() {
        assert_eq!(canonical_json(&json!({ "after": null, "before": 1 })), "{\"after\":null,\"before\":1}");
    }
}
